src/libc/gen/clock_gettime.3
.\"
.\" Copyright (c) 1980, 1991, 1993
.\" The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd January 26, 2016
.Dt CLOCK_GETTIME 3
.Os
.Sh NAME
.Nm clock_gettime ,
.Nm clock_settime ,
.Nm clock_getres ,
.Nm clock_gettime_nsec_np
.Nd get/set date and time
.Sh SYNOPSIS
.Fd #include <time.h>
.Ft int
.Fn clock_gettime "clockid_t clock_id" "struct timespec *tp"
.Ft int
.Fn clock_settime "clockid_t clock_id" "const struct timespec *tp"
.Ft int
.Fn clock_getres "clockid_t clock_id" "struct timespec *tp"
.Ft uint64_t
.Fn clock_gettime_nsec_np "clockid_t clock_id"
.Sh DESCRIPTION
The
.Fn clock_gettime
and
.Fn clock_settime
functions
allow the calling process to retrieve or set the value used by a clock
which is specified by
.Fa clock_id .
.Pp
.Fa clock_id
can be a value from
one of 8 predefined values:
.Bl -tag -width Er
.It Dv CLOCK_REALTIME
the system's real time (i.e. wall time) clock, expressed as the amount of time
since the Epoch.
This is the same as the value returned by
.Xr gettimeofday 2 .
.It Dv CLOCK_MONOTONIC
clock that increments monotonically, tracking the time since an arbitrary
point, and will continue to increment while the system is asleep.
.It Dv CLOCK_MONOTONIC_RAW
clock that increments monotonically, tracking the time since an arbitrary point
like CLOCK_MONOTONIC.  However, this clock is unaffected by frequency or time
adjustments.  It should not be compared to other system time sources.
.It Dv CLOCK_MONOTONIC_RAW_APPROX
like CLOCK_MONOTONIC_RAW, but reads a value cached by the system at context
switch.  This can be read faster, but at a loss of accuracy as it may return
values that are milliseconds old.
.It Dv CLOCK_UPTIME_RAW
clock that increments monotonically, in the same manner as
.Dv CLOCK_MONOTONIC_RAW,
but that does not increment while the system is asleep.
The returned value is identical to the result of
.Fn mach_absolute_time
after the appropriate mach_timebase conversion is applied.
.It Dv CLOCK_UPTIME_RAW_APPROX
like CLOCK_UPTIME_RAW, but reads a value cached by the system at context
switch.  This can be read faster, but at a loss of accuracy as it may return
values that are milliseconds old.
.It Dv CLOCK_PROCESS_CPUTIME_ID
clock that tracks the amount of CPU (in user- or kernel-mode) used by the
calling process.
.It Dv CLOCK_THREAD_CPUTIME_ID
clock that tracks the amount of CPU (in user- or kernel-mode) used by the
calling thread.
.El
.Pp
The structure pointed to by
.Fa tp
is defined in
.In sys/time.h
as:
.Bd -literal -offset indent
struct timespec {
    time_t  tv_sec;     /* seconds */
    long    tv_nsec;    /* and nanoseconds */
};
.Ed
.Pp
Only the
.Dv CLOCK_REALTIME
clock can be set, and only the superuser may do so.
.Pp
The resolution of a clock is returned by the
.Fn clock_getres
call.
This value is placed in a (non-null)
.Fa *tp .
This value may be smaller than the actual precision of the underlying clock,
but represents a lower bound on the resolution.
.Pp
As a non-portable extension, the
.Fn clock_gettime_nsec_np
function will return the clock value in 64-bit nanoseconds.
.Sh RETURN VALUES
A 0 return value indicates that the call succeeded.
A \-1 return value indicates an error occurred, and in this
case an error code is stored into the global variable
.Va errno .
For
.Fn clock_gettime_nsec_np
a return value of non-0 indicates success.  A 0 return value indicates an error
occurred and an error code is stored in
.Va errno .
.Sh ERRORS
.Fn clock_gettime ,
.Fn clock_settime ,
.Fn clock_getres ,
and
.Fn clock_gettime_nsec_np
will fail if:
.Bl -tag -width Er
.It Bq Er EINVAL
.Fa clock_id
is not a valid value.
.It Bq Er EFAULT
The
.Fa tp
argument address referenced invalid memory.
.El
.Pp
In addition,
.Fn clock_settime
may return the following errors:
.Bl -tag -width Er
.It Bq Er EPERM
A user other than the superuser attempted to set the time.
.It Bq Er EINVAL
.Fa clock_id
specifies a clock that isn't settable,
.Fa tp
specifies a nanosecond value less than zero or greater than 1000 million,
or a value outside the range of the specified clock.
.El
.Sh SEE ALSO
.Xr date 1 ,
.Xr getitimer 2 ,
.Xr gettimeofday 2 ,
.Sh HISTORY
These functions first appeared
in
Mac OSX 10.12
.Sh STANDARDS
The
.Fn clock_gettime ,
.Fn clock_settime ,
and
.Fn clock_getres
system calls conform to
.St -p1003.1b-93 .
.Fn cleck_gettime_nsec_np
is a non-portable Darwin extension.
The clock IDs
.Fa CLOCK_MONOTONIC_RAW
and
.Fa CLOCK_UPTIME_RAW
are extensions to the POSIX interface.


src/libc/gen/clock_gettime.c
/*
 * Copyright (c) 2016 Apple Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */

#include <errno.h>
#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/sysctl.h>
#include <mach/mach_error.h>
#include <mach/mach_time.h>
#include <os/base_private.h>

extern uint64_t __thread_selfusage(void);

#define timeval2nsec(tv) (tv.tv_sec * NSEC_PER_SEC + tv.tv_usec * NSEC_PER_USEC)

static uint64_t
_boottime_fallback_usec(void)
{
    struct timeval tv;
    size_t len = sizeof(tv);
    int ret = sysctlbyname("kern.boottime", &tv, &len, NULL, 0);
    if (ret == -1) return 0;
    return (uint64_t)tv.tv_sec * USEC_PER_SEC + (uint64_t)tv.tv_usec;
}

static int
_mach_boottime_usec(uint64_t *boottime, struct timeval *realtime)
{
    uint64_t bt1 = 0, bt2 = 0;
    int ret;
    do {
        bt1 = mach_boottime_usec();
        if (os_slowpath(bt1 == 0)) bt1 = _boottime_fallback_usec();

        atomic_thread_fence(memory_order_seq_cst);

        ret = gettimeofday(realtime, NULL);
        if (ret != 0) return ret;

        atomic_thread_fence(memory_order_seq_cst);

        bt2 = mach_boottime_usec();
        if (os_slowpath(bt2 == 0)) bt2 = _boottime_fallback_usec();
    } while (os_slowpath(bt1 != bt2));
    *boottime = bt1;
    return 0;
}

uint64_t
clock_gettime_nsec_np(clockid_t clock_id)
{
    switch(clock_id){
    case CLOCK_REALTIME: {
        struct timeval tv;
        int ret = gettimeofday(&tv, NULL);
        if (ret) return 0;
        return timeval2nsec(tv);
    }
    case CLOCK_MONOTONIC: {
        struct timeval tv;
        uint64_t boottime;
        int ret = _mach_boottime_usec(&boottime, &tv);
        if (ret) return 0;
        boottime *= NSEC_PER_USEC;
        return timeval2nsec(tv) - boottime;
    }
    case CLOCK_PROCESS_CPUTIME_ID: {
        struct rusage ru;
        int ret = getrusage(RUSAGE_SELF, &ru);
        if (ret) return 0;
        return timeval2nsec(ru.ru_utime) + timeval2nsec(ru.ru_stime);
    }
    default:
        // calls that use mach_absolute_time units fall through into a common path
        break;
    }

    // Mach Absolute Time unit-based calls
    mach_timebase_info_data_t tb_info;
    if (mach_timebase_info(&tb_info)) return 0;
    uint64_t mach_time;

    switch(clock_id){
    case CLOCK_MONOTONIC_RAW:
        mach_time = mach_continuous_time();
        break;
    case CLOCK_MONOTONIC_RAW_APPROX:
        mach_time = mach_continuous_approximate_time();
        break;
    case CLOCK_UPTIME_RAW:
        mach_time = mach_absolute_time();
        break;
    case CLOCK_UPTIME_RAW_APPROX:
        mach_time = mach_approximate_time();
        break;
    case CLOCK_THREAD_CPUTIME_ID:
        mach_time = __thread_selfusage();
        break;
    default:
        errno = EINVAL;
        return 0;
    }

    return (mach_time * tb_info.numer) / tb_info.denom;
}

int
clock_gettime(clockid_t clk_id, struct timespec *tp)
{
    switch(clk_id){
    case CLOCK_REALTIME: {
        struct timeval tv;
        int ret = gettimeofday(&tv, NULL);
        TIMEVAL_TO_TIMESPEC(&tv, tp);
        return ret;
    }
    case CLOCK_MONOTONIC: {
        struct timeval tv;
        uint64_t boottime_usec;
        int ret = _mach_boottime_usec(&boottime_usec, &tv);
        struct timeval boottime = {
            .tv_sec = boottime_usec / USEC_PER_SEC,
            .tv_usec = boottime_usec % USEC_PER_SEC
        };
        timersub(&tv, &boottime, &tv);
        TIMEVAL_TO_TIMESPEC(&tv, tp);
        return ret;
    }
    case CLOCK_PROCESS_CPUTIME_ID: {
        struct rusage ru;
        int ret = getrusage(RUSAGE_SELF, &ru);
        timeradd(&ru.ru_utime, &ru.ru_stime, &ru.ru_utime);
        TIMEVAL_TO_TIMESPEC(&ru.ru_utime, tp);
        return ret;
    }
    case CLOCK_MONOTONIC_RAW:
    case CLOCK_MONOTONIC_RAW_APPROX:
    case CLOCK_UPTIME_RAW:
    case CLOCK_UPTIME_RAW_APPROX:
    case CLOCK_THREAD_CPUTIME_ID: {
        uint64_t ns = clock_gettime_nsec_np(clk_id);
        if (!ns) return -1;

        tp->tv_sec = ns/NSEC_PER_SEC;
        tp->tv_nsec = ns % NSEC_PER_SEC;
        return 0;
    }
    default:
        errno = EINVAL;
        return -1;
    }
}

int
clock_getres(clockid_t clk_id, struct timespec *res)
{
    switch(clk_id){
    case CLOCK_REALTIME:
    case CLOCK_MONOTONIC:
    case CLOCK_PROCESS_CPUTIME_ID:
        res->tv_nsec = NSEC_PER_USEC;
        res->tv_sec = 0;
        return 0;

    case CLOCK_MONOTONIC_RAW:
    case CLOCK_MONOTONIC_RAW_APPROX:
    case CLOCK_UPTIME_RAW:
    case CLOCK_UPTIME_RAW_APPROX:
    case CLOCK_THREAD_CPUTIME_ID: {
        mach_timebase_info_data_t tb_info;
        if (mach_timebase_info(&tb_info)){
            return -1;
        }
        res->tv_nsec = tb_info.numer / tb_info.denom + (tb_info.numer % tb_info.denom != 0);
        res->tv_sec = 0;
        return 0;
    }

    default:
        errno = EINVAL;
        return -1;
    }
}

int
clock_settime(clockid_t clk_id, const struct timespec *tp)
{
    switch(clk_id){
    case CLOCK_REALTIME: {
        struct timeval tv = {
            .tv_sec = (time_t)tp->tv_sec,
            .tv_usec = (suseconds_t)(tp->tv_nsec / NSEC_PER_USEC)
        };
        return settimeofday(&tv, NULL);
    }
    default:
        errno = EINVAL;
        return -1;
    }
}
